### PR TITLE
citra_qt, video_core: Screenshot functionality

### LINKED ISSUE #3121: Screenshot Function for Citra
_I was [instructed by a fellow member](https://community.citra-emu.org/t/any-potential-for-a-screenshot-option/1706/6?u=minimii) to start an issue here._
A built-in screenshot function in Citra would be great to have, my main reason for needing it en lieu of the usual "Windows" way (using PrntScr) is because I run some games at native resolution but with a stretched window. Doing this causes some blurriness to show, which would be captured if I use PrntScr and I don't want that... What I want is for screenshots to have the game's internal resolution, regardless of the user's window size/layout.

_**Example:** Native resolution in a non-stretched window vs. in a stretched window (enlarge)_

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
Closes #3121

Allows capturing screenshot at the current internal resolution (native for software renderer), but a setting is available to capture it in other resolutions (the internal resolution won't be changed, instead the image captured is upscaled/downscaled). The screenshot is saved to a single PNG in the current layout.

Examples of captured images can be found in the #screenshots-videos channel in Citra discord (they were captured with an older version of the code, but I have confirmed the result is the same). This is [one](https://cdn.discordapp.com/attachments/472575481179537418/484372018179997706/screenshot2.zip) of them.

A menu action is added to the Movie menu (I believe this is the most suitable place) as well as a hotkey. The hotkey is bound to Ctrl+S by default.

I have just started to learn GL stuff, and there may be easier ways to achieve what I'm doing in renderer_opengl.cpp. If that's the case, please point it out. Thanks!

Screenshot:
image

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4164)

### CODE REVIEW

**Comment by FearlessTobi** [comment]
As for every UI change, can you please provide a screenshot for every changed dialog/setting?

**Comment by zhaowenlan1779** [comment]
@FearlessTobi  huh.. I didn't think it was necessary as it is just a simple and common menu action. But fine :P added it to the description.

**Comment by j-selby** [comment]
Can I suggest renaming the Movie option to something more general? "Capture"?

**Comment by zhaowenlan1779** [comment]
@j-selby "Capture" does not sound so good... I cannot think of a better one right now.
Also, the setting does not exist in the UI as I think it would cause confusion if the image captured became pixelated due to upscaling. Normally the image captured would just be at the internal resolution.

**Comment by B3n30** [comment]
TBH I don't think screenshot should be a sub-option of movie. "Movies" in emulators are known to be input-recordings. And IMO screenshots have not very much in common with such movies.

Maybe do what @jroweboy suggested. Make a option tools and inside tools there is the movie submenu and the screenshot function?

**Comment by valentinvanelslande** [comment]
Issues:
- Game is not resumed if you cancel the dialog.
- Colors issue with FBI (screenshot from my latest commit + this pr + some other changes): 
issue

**Comment by zhaowenlan1779** [comment]
Color looks correct to me. I believe the blue one was correct?
image

**Comment by BreadFish64** [comment]
The blue is correct yes

**Comment by zhaowenlan1779** [comment]
Fixed the first issue.

**Comment by wwylele** [comment]
The issue @valentinvanelslande has is probably with the image viewer on linux which does some "color correction". I had the same issue on ubuntu.

**Comment by B3n30** [comment]
I didn't make a full review but doesn't that introduce some potential race conditions between the qt_thread and the emulation (video-core) thread?

**Comment by wwylele** [comment]
@B3n30 qt thread only sends a request to emu thread (done by an atomic bit) and the real capture-and-save operation is done entirely in the emu thread.

File: src/video_core/video_core.cpp [hunk 1]
```diff
@@ -48,4 +54,27 @@ void Shutdown() {
     LOG_DEBUG(Render, "shutdown OK");
 }
 
+void RequestScreenshot(void* data, std::function<void()> callback,
+                       const Layout::FramebufferLayout& layout) {
+    if (g_renderer_screenshot_requested) {
+        LOG_ERROR(Render, "A screenshot is already requested or in progress, ignoring the request");
```
Comment (baka0815) [inline src/video_core/video_core.cpp]: Does this need to be an error? Isn't a warning more appropriate here?

File: src/video_core/renderer_opengl/renderer_opengl.cpp [hunk 2]
```diff
@@ -138,7 +138,39 @@ void RendererOpenGL::SwapBuffers() {
...
+        Layout::FramebufferLayout layout{VideoCore::g_screenshot_framebuffer_layout};
+
+        GLuint renderbuffer;
+        glGenRenderbuffers(1, &renderbuffer);
+        glBindRenderbuffer(GL_RENDERBUFFER, renderbuffer);
+        glRenderbufferStorage(GL_RENDERBUFFER, GL_RGB8, layout.width, layout.height);
+        glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER,
+                                  renderbuffer);
+
+        DrawScreens(layout);
```
Comment (baka0815) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Doesn't this skip a frame if a screenshot should be taken?

Could we just "copy" this layout to ``render_window.GetFramebufferLayout()``?
Or am I missing something?

Comment (zhaowenlan1779) [reply src/video_core/renderer_opengl/renderer_opengl.cpp]: The same frame is drawn to the screen as well a few lines later, outside of the "if".

**Comment by Leo626** [comment]
I noticed Ctrl + S is the same hotkey for Toggle Status Bar.

File: src/video_core/video_core.cpp (hunk 1 above)
Comment (zhaowenlan1779) [reply src/video_core/video_core.cpp]: Not sure about that. IMO this should be an error, as a user request is ignored.

**Comment by zhaowenlan1779** [comment]
@Leo626 Thanks for reporting. Will change the hotkey to something else...

**Comment by zhaowenlan1779** [comment]
This PR is now ready for review.

**Comment by zhaowenlan1779** [comment]
I'll merge in 24 hours if no more comments

**Comment by zhaowenlan1779** [comment]
Oh wait, I still need to change the hotkey.

**Comment by RocketRobz** [comment]
A minor request: How about an option to set a resolution specific to screenshot captures?
For me, setting the internal resolution to 10x makes the game run a bit slow, and setting it just for hi-res screenshots, and then back to the previous setting for regular usage, can get pretty tiresome.

I'm thinking of having the resolution list shown when hovering over `Capture Screenshot`.

**Comment by zhaowenlan1779** [comment]
there's already a config entry for it. I can add that to the UI if it's needed.

**Comment by wwylele** [comment]
Will merge this soon if no more comments

**Comment by BreadFish64** [comment]
I think the screenshot capture resolution warrants a UI option

**Comment by zhaowenlan1779** [comment]
yeah, will add that tomorrow

**Comment by zhaowenlan1779** [comment]
Added resolution selection and changed hotkey to CTRL+P.

**Comment by Narugakuruga** [comment]
This option seems to only change capture resolution but not rendering resolution.

**Comment by zhaowenlan1779** [comment]
@Narugakuruga ah, that's right. I'll fix it soon

**Comment by zhaowenlan1779** [comment]
@RocketRobz @Narugakuruga 
Sorry, due to technical reasons this is actually harder than it sounds and would require dirty hacks currently. I have dropped the last commit. People are welcome to reinvent it in the future when this becomes easier.

**Comment by Narugakuruga** [comment]
I understood, this option is actually not really necessary, people can change rendering resolution in graphics configure manually

**Pull request merged**